fetch/fetch.go

```diff
@@ -130,17 +105,92 @@ func randomPeer(peers []p2p.Peer) p2p.Peer {
 	return peers[rand.Intn(len(peers))]
 }
 
+// Option is a type to configure a fetcher.
+type Option func(*Fetch)
+
+// WithConfig configures the config for the fetcher.
+func WithConfig(c Config) Option {
+	return func(f *Fetch) {
+		f.cfg = c
+	}
+}
+
+// WithLogger configures logger for the fetcher.
+func WithLogger(log log.Log) Option {
+	return func(f *Fetch) {
+		f.logger = log
+	}
+}
+
+// WithATXHandlers configures the ATX handler of the fetcher.
+func WithATXHandlers(h atxHandler) Option {
+	return func(f *Fetch) {
+		f.atxHandler = h
+	}
+}
+
+// WithBallotHandlers configures the Ballot handler of the fetcher.
+func WithBallotHandlers(h ballotHandler) Option {
+	return func(f *Fetch) {
+		f.ballotHandler = h
+	}
+}
+
+// WithBlockHandlers configures the Block handler of the fetcher.
+func WithBlockHandlers(h blockHandler) Option {
+	return func(f *Fetch) {
+		f.blockHandler = h
+	}
+}
+
+// WithProposalHandlers configures the Proposal handler of the fetcher.
+func WithProposalHandlers(h proposalHandler) Option {
+	return func(f *Fetch) {
+		f.proposalHandler = h
+	}
+}
+
+// WithTXHandlers configures the TX handler of the fetcher.
+func WithTXHandlers(h txHandler) Option {
+	return func(f *Fetch) {
+		f.txHandler = h
+	}
+}
+
+// WithPoetHandlers configures the PoET handler of the fetcher.
+func WithPoetHandlers(h poetHandler) Option {
+	return func(f *Fetch) {
+		f.poetHandler = h
+	}
+}
+
+func withServers(s map[string]requester) Option {
+	return func(f *Fetch) {
+		f.servers = s
+	}
+}
+
+func withHost(h host) Option {
+	return func(f *Fetch) {
+		f.host = h
+	}
+}
+
 // Fetch is the main struct that contains network peers and logic to batch and dispatch hash fetch requests.
 type Fetch struct {
-	cfg     Config
-	log     log.Log
-	eg      errgroup.Group
-	bs      *datastore.BlobStore
-	host    host
-	atxSrv  server.Requestor
-	lyrSrv  server.Requestor
-	opnSrv  server.Requestor
-	hashSrv server.Requestor
+	cfg    Config
+	logger log.Log
+	eg     errgroup.Group
+	bs     *datastore.BlobStore
+	host   host
+
+	servers         map[string]requester
+	poetHandler     poetHandler
+	atxHandler      atxHandler
+	ballotHandler   ballotHandler
+	blockHandler    blockHandler
+	proposalHandler proposalHandler
+	txHandler       txHandler
```

Comment:
Handlers are such generic names; would `...Hook` also work?

Reply:
hmmm.... i don't see much difference btwn Handler and Hook. and the instances passed in are called handlers in struct.
if we are just passing a func then i'd be more inclined to change.

Reply:
Besides `txHandler` all handlers are essentially just callback functions. Take `poetHandler` as an example, it could be `onPoetReceived` instead:

```golang
type onPoetReceived (data []byte) error
```

and then be passed in like this:

```golang
// node.go
fetch.WithPoetHook(poetDb.ValidateAndStoreMsg)

// fetch_test.go
WithPoetHook(tf.mPoetH.ValidateAndStoreMsg),
```

I feel like this gives a better indication on what is handled by the handlers.

Reply:
two considerations:
- now we have function ptr for all data types other than TXs
- it makes the testing setup a bit more difficult. with handler instance i can use gomock. with func ptr i have to create individual functions to pass to test.

i don't see why function ptr is better in this case.

Reply:
> now we have function ptr for all data types other than TXs

`txHandler` is arguably `onBlockTxReceived` and `onProposalTxReceived` and could be two hooks.

> it makes the testing setup a bit more difficult. with handler instance i can use gomock. with func ptr i have to create individual functions to pass to test.

Interesting, I did not know that `gomock` cannot mock functions. Theoretically it should be able to do so, this is how the "generated" code would look like.

```golang
package mocks

import (
	reflect "reflect"

	gomock "github.com/golang/mock/gomock"
)

// MockOnPoetReceived is a mock of onPoetReceived function.
type MockOnPoetReceived struct {
	ctrl     *gomock.Controller
	recorder *MockOnPoetReceivedMockRecorder
}

// MockOnPoetReceivedMockRecorder is the mock recorder for MockOnPoetReceived.
type MockOnPoetReceivedMockRecorder struct {
	mock *MockOnPoetReceived
}

// NewMockOnPoetReceived creates a new mock instance.
func NewMockOnPoetReceived(ctrl *gomock.Controller) *MockOnPoetReceived {
	mock := &MockOnPoetReceived{ctrl: ctrl}
	mock.recorder = &MockOnPoetReceivedMockRecorder{mock}
	return mock
}

// EXPECT returns an object that allows the caller to indicate expected use.
func (m *MockOnPoetReceived) EXPECT() *MockOnPoetReceivedMockRecorder {
	return m.recorder
}

// Call mocks base method.
func (m *MockOnPoetReceived) Call(data []byte) error {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "Call", data)
	ret0, _ := ret[0].(error)
	return ret0
}

// Call indicates an expected call of Call.
func (mr *MockOnPoetReceivedMockRecorder) Call(data interface{}) *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Call", reflect.TypeOf((*MockOnPoetReceived)(nil).Call), data)
}
```

And this is how it would be used:

```golang
ctrl := gomock.NewController(tb)
mPoetH := mocks.NewMockOnPoetReceived(ctrl)
fetch := NewFetch(datastore.NewCachedDB(sql.InMemory(), lg), tf.mMesh, nil,
		// some other functional options
		WithPoetHook(mPoetH.Call),
)

mPoetH.EXPECT().Call(data).Return(nil).Times(1)
```

Reply:
Since this is not possible at the moment, let's leave it like it is.

Reply:
maybe you should contribute this to gomock then! mocking functions will be super useful.